### PR TITLE
Feature/mico#562 Restructure service deployment information database schema

### LINKED ISSUE #556: Fix update deployment information
Endpoint `updateServiceDeploymentInformation` does not update labels. Response is ok, but a subsequent GET request returns an empty list of labels.

PUT Request:
```
curl -X PUT \
  http://localhost:8080/applications/hello/v0.0.1/deploymentInformation/hello \
  -d '{
    "replicas": 2,
    "minReadySecondsBeforeMarkedAvailable": 5,
    "labels": [
        {
            "key": "value1",
            "value": "value1"
        }
    ],
    "imagePullPolicy": "NEVER",
    "restartPolicy": "NEVER"
}'
```

Response:
```
{
    "replicas": 2,
    "minReadySecondsBeforeMarkedAvailable": 5,
    "labels": [
        {
            "key": "value1",
            "value": "value1"
        }
    ],
    "imagePullPolicy": "NEVER",
    "restartPolicy": "NEVER",
    "_links": {
        "self": {
            "href": "http://localhost:8080/applications/hello/v0.0.1/deploymentInformation/hello"
        }
    }
}
```

GET Request:
```
curl -X GET \
  http://localhost:8080/applications/hello/v0.0.1/deploymentInformation/hello
```

Response:
```
{
    "replicas": 2,
    "minReadySecondsBeforeMarkedAvailable": 5,
    "labels": [],
    "imagePullPolicy": "NEVER",
    "restartPolicy": "NEVER",
    "_links": {
        "self": {
            "href": "http://localhost:8080/applications/hello/v0.0.1/deploymentInformation/hello"
        }
    }
}
```

### LINKED ISSUE #561: Move `deleteInterfaceOfServiceByName()` from `MicoServiceRepository` to `MicoServiceInterfaceRepository`
The above mention function deletes an instance of `MicoServiceInterface` and thus should be located in the corresponding repository (`MicoServiceInterfacerReposiory`).

### LINKED ISSUE #562: Restructure domain model representation of service deployment information.
Currently, a `MicoServiceDeploymentInfo` (domain) object is represented as a `Relationship` in the database.

Since Neo4j does not allow composite types such as lists and maps as properties on relationships, and we need, to store environment variables and labels as part of the deployment information of a `MicoService`, `MicoServiceDeploymentInformation` needs to be stored as a `Node` in the database. 

The new graph schema will look like this:
sdi-1

Considering `MicoService`s that depend on other `MicoService`s it is possible for a `MicoApplication` to specify deployment information also for the dependencies of the included `MicoService`s (see below for an example scenario).

sdi-2

The complete graph schema (including properties of `MicoServiceDeploymentInfo`) will look something like this:

sdi-3

Fixes #556.

### PR DESCRIPTION
**Pull request by JakobBenz**
- [x] Ensure that the commit message is [a good commit message](https://github.com/joelparkerhenderson/git_commit_message)
- [x] Ensure to use auto format in **all** files
- [x] Tests created for changes or: test created / adjusted.

---

## Short Description

Currently, a MicoServiceDeploymentInfo (domain) object is represented as a Relationship in the database. Since Neo4j does not allow composite types such as lists and maps as properties on relationships, and we need, to store environment variables and labels as part of the deployment information of a `MicoService`, `MicoServiceDeploymentInfo` needs to be stored as a separate node entity in the database.

For the updated graph schema see below. Note that this pull request does not include `MicoEnvironmentVariable` (see #525) and `KubernetesDeploymentInfo` (see #439) colored in light grey.

sdi-3

## Resolving Issue/ Feature

Fixes #556.
Resolves #561.
Resolves #562.
Closes #525.

### CODE REVIEW

File: mico-core/src/main/java/io/github/ust/mico/core/resource/ApplicationResource.java [hunk 1]
```diff
@@ -299,33 +333,37 @@
...
+    	MicoServiceDeploymentInfo updatedServiceDeploymentInfo = serviceDeploymentInfoRepository.save(serviceDeploymentInfoOptional.get().applyValuesFrom(serviceDeploymentInfoDTO));
+    	// In case some labels have been removed from the list of labels in the service deployment information,
+    	// the standard save() function of the service deployment information repository will not delete those
+    	// "tangling" (without relationships) labels (nodes), hence the manual clean up.
+    	// TODO: Alternative: use delete + insert instead of update. Decide whether to do so or not.
+    	labelRepository.cleanUp();
+    	// Same goes for the environment variables, ...
+    	// TODO: Clean up environment variables.
+    	// ... and the Kubernetes deployment info.
+    	// TODO: Clean up Kubernetes deployment info.
```
Comment (davidkopp) [inline mico-core/src/main/java/io/github/ust/mico/core/resource/ApplicationResource.java]: Resolve ToDos before merge

File: mico-core/src/main/java/io/github/ust/mico/core/resource/ApplicationResource.java [hunk 2]
```diff
@@ -86,252 +93,288 @@
...
+        
+        // In order to copy the application along with all service deployment information nodes
+        // it provides and the nodes the service deployment information node(s) is/are connected to,
+        // we need to set the id of all those nodes to null. That way, Neo4j will create new entities
+        // instead of updating the existing ones.
+        application.setId(null);
+        application.getServiceDeploymentInfos().forEach(sdi -> sdi.setId(null));
+        application.getServiceDeploymentInfos().forEach(sdi -> sdi.getLabels().forEach(label -> label.setId(null)));
+        application.getServiceDeploymentInfos().forEach(sdi -> sdi.getEnvironmentVariables().forEach(envVar -> envVar.setId(null)));
+        // TODO: Also clear ids for Kubernetes deployment information.
```
Comment (davidkopp) [inline mico-core/src/main/java/io/github/ust/mico/core/resource/ApplicationResource.java]: ToDo must be resolved before merge

File: mico-core/src/main/java/io/github/ust/mico/core/resource/ApplicationResource.java (hunk 1 above)
Comment (JakobBenz) [reply mico-core/src/main/java/io/github/ust/mico/core/resource/ApplicationResource.java]: Done.

File: mico-core/src/main/java/io/github/ust/mico/core/resource/ApplicationResource.java (hunk 2 above)
Comment (JakobBenz) [reply mico-core/src/main/java/io/github/ust/mico/core/resource/ApplicationResource.java]: Done.

**Pull request merged**